### PR TITLE
Add subscription-level rule manager

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Addresses https://github.com/Azure/azure-sdk-for-net/issues/25523

Re-implements the API that we had been using before pulling this from GA. It will need to be re-reviewed.

Marking as Do Not Merge as I don't want to impact the GA release planned for next week.

### CODE REVIEW

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpRuleManager.cs [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+        /// <value>
+        /// <c>true</c> if the rule manager is closed; otherwise, <c>false</c>.
+        /// </value>
+        public override bool IsClosed => _closed;
+
+        private readonly FaultTolerantAmqpObject<RequestResponseAmqpLink> _managementLink;
+
+        static AmqpRuleManager()
+        {
+            AmqpCodec.RegisterKnownTypes(AmqpTrueRuleFilterCodec.Name, AmqpTrueRuleFilterCodec.Code, () => new AmqpTrueRuleFilterCodec());
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpRuleManager.cs:61]: Interesting.   Never seen this before.   I'd be very interested in learning the details at some point, if you have time.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/RuleManager/ServiceBusRuleManager.cs [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+                    ruleName,
+                    cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception exception)
+            {
+                ServiceBusEventSource.Log.RemoveRuleException(Identifier, exception.ToString());
+                throw;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/RuleManager/ServiceBusRuleManager.cs:182]: Do we really want to check cancellation here?   The operation is already complete.

If we do decide to keep it, if this line throws, you'll get a "Remove Rule Start" log event without a matching error/complete.   Is that intended?

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/RuleManager/RuleManagerLiveTests.cs [hunk 3]
```diff
@@ -0,0 +1,589 @@
...
+using System.Threading.Tasks;
+using Azure.Messaging.ServiceBus.Administration;
+using NUnit.Framework;
+
+namespace Azure.Messaging.ServiceBus.Tests.RuleManager
+{
+    public class RuleManagerLiveTests : ServiceBusLiveTestBase
+    {
+        [Test]
+        public async Task AddGetAndRemoveRules()
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/tests/RuleManager/RuleManagerLiveTests.cs:18]: That's a pretty large amount of functionality for a single test to be responsible for.   Should we consider breaking this out into three tests to isolate the Get, Add, and Remove scopes?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/RuleManager/ServiceBusRuleManager.cs [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+        }
+
+        /// <summary>
+        /// Get all rules associated with the subscription.
+        /// </summary>
+        ///
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        /// <returns>Returns a list of rules description</returns>
+        public virtual async Task<IList<RuleProperties>> GetRulesAsync(CancellationToken cancellationToken = default)
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/RuleManager/ServiceBusRuleManager.cs:193]: Should we consider going with `IReadOnlyList` for these?

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/RuleManager/RuleManagerLiveTests.cs [hunk 5]
```diff
@@ -0,0 +1,589 @@
...
+                        ReplyTo = "replyTo",
+                        ReplyToSessionId = "replyToSessionId",
+                        SessionId = "sessionId",
+                        To = "to"
+                    },
+                    Action = new SqlRuleAction("Set CorrelationId = 'newValue'")
+                };
+                await ruleManager.AddRuleAsync(ruleOptions);
+
+                rules = (await ruleManager.GetRulesAsync()).ToList();
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/tests/RuleManager/RuleManagerLiveTests.cs:56]: `GetRulesAsync` returns a list by contract.   Do we need to `ToList()` here for some reason?

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/RuleManager/RuleManagerLiveTests.cs [hunk 6]
```diff
@@ -0,0 +1,589 @@
...
+                Assert.AreEqual("sessionId", correlationRuleFilter.SessionId);
+                Assert.AreEqual("to", correlationRuleFilter.To);
+                Assert.NotNull(correlationRuleFilter.ApplicationProperties);
+                Assert.AreEqual("value1", correlationRuleFilter.ApplicationProperties["key1"]);
+
+                await ruleManager.RemoveRuleAsync(RuleProperties.DefaultRuleName);
+                await ruleManager.RemoveRuleAsync(sqlRuleName);
+                await ruleManager.RemoveRuleAsync(correlationRuleName);
+                rules = (await ruleManager.GetRulesAsync()).ToList();
+                Assert.AreEqual(0, rules.Count());
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/tests/RuleManager/RuleManagerLiveTests.cs:88]:
```suggestion
                Assert.AreEqual(0, rules.Count);
```

Not impactful, but we know we have a list here.

**Review by jsquire — APPROVED** [review]

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/RuleManager/RuleManagerLiveTests.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/RuleManager/RuleManagerLiveTests.cs:18]: Well each of them will still need to do the Add step, so I think it is probably okay to combine them.

**Pull request merged**